lib/srv/regular/proxy.go

```diff
@@ -279,13 +280,52 @@ func (t *proxySubsys) proxyToHost(ctx context.Context, ch ssh.Channel, clientSrc
 	if teleportVersion != "" && utils.CheckVersion(teleportVersion, utils.MinIPPropagationVersion) != nil {
 		t.doHandshake(ctx, clientSrcAddr, ch, conn)
 	}
+
+	checkedReader := &checkedPrefixReader{
+		ReadWriteCloser: ch,
+		// SSH connection MUST start with "SSH-2.0" bytes according to https://datatracker.ietf.org/doc/html/rfc4253#section-4.2
+		requiredPrefix: []byte("SSH-2.0"),
+	}
+
 	go func() {
-		t.close(utils.ProxyConn(ctx, ch, conn))
+		t.close(utils.ProxyConn(ctx, checkedReader, conn))
 	}()
 
 	return nil
 }
 
+// checkedPrefixReader checks that read data has the specified prefix.
+type checkedPrefixReader struct {
+	io.ReadWriteCloser
+	requiredPrefix []byte
+
+	// upstreamReader reads from the underlying reader.
+	upstreamReader io.Reader
+}
+
+func (c *checkedPrefixReader) Read(b []byte) (int, error) {
+	// connection was already checked, forward upstream:
+	if c.upstreamReader != nil {
+		return c.upstreamReader.Read(b)
+	}
+
+	// make sure first bytes are equal to the required prefix
+	reader := bufio.NewReader(c.ReadWriteCloser)
```

Comment:
This will make a new bufio reader on top of the inner RWC on every read, reading (in general) more than what will be returned to the user, then discarding the data.